src/cast_core/src/RunCommand.cpp

```diff
@@ -208,13 +208,110 @@ class DatabaseOperation {
+        static constexpr std::string_view kSharded = "sharded";
+        static constexpr std::string_view kReplicaSet = "replica_set";
+
+        static std::string getValidOptionsListAsString() {
+            std::stringstream listString;
+            listString << kStandalone << ", " << kSharded << ", " << kReplicaSet;
+            return listString.str();
+        }
+
+        static bool isValid(std::string_view instanceType) {
+            return instanceType == kStandalone || instanceType == kSharded
+                    || instanceType == kReplicaSet;
+        }
+
+        static bool isValidList(const std::vector<std::string>& options) {
+            for(auto &option : options) {
+                if(!isValid(option)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+
+    InstanceTypeFilter(PhaseContext &context, mongocxx::pool::entry& client) {
+        try {
+            _onlyRunInInstancesContext = context.getPlural<std::string>(
+                "OnlyRunInInstance", "OnlyRunInInstances", [&](const Node& node) {
+                    return node.to<std::string>();
+            });
+        } catch(...) {
+            // Exception might be due to keys not found or other errors. If its due to keys not
+            // found ignore the expception. Otherwise rethrow.
+            if(context["OnlyRunInInstance"] || context["OnlyRunInInstance"]) {
+                std::rethrow_exception(std::current_exception());
+            }
+        }
+
+        if(_onlyRunInInstancesContext.size()) {
+            if(!Options::isValidList(_onlyRunInInstancesContext)) {
+                std::stringstream errorMsg;
+                errorMsg << "OnlyRunInInstance or OnlyRunInInstances valid values are: "
+                    << Options::getValidOptionsListAsString();
+                throw InvalidConfigurationException(errorMsg.str());
+            }
+
+            _adminDB = (*client)["admin"];
+        }
+    }
+
+    // Indicates if operations should be run. If there is no OnlyRunInInstance configuration or one
+    // of the specified options matches the client instance type, returns true. Otherwise returns
+    // false.
+    bool shouldRun() {
+        if(_type.length() == 0 && _onlyRunInInstancesContext.size()) {
+            _type = getInstanceType(_adminDB);
+            _typeFoundInConfig = false;
+            for(const auto &configInstanceType : _onlyRunInInstancesContext) {
+                if(_type == configInstanceType) {
+                    _typeFoundInConfig = true;
+                    break;
+                }
+            }
+        }
+        return _onlyRunInInstancesContext.size() == 0 || _typeFoundInConfig;
+    }
+
+    // The output of the "hello" command is parsed to check if we are running in a Mongos, a replica
+    // set, or a standalone instance.
+    static std::string_view getInstanceType(mongocxx::database &db) {
```

Comment:
Would it be possible to make use of the toplogy visitor: https://github.com/mongodb/genny/blob/5a525da35c8153ae59bc20af190b0dd454ac1975/src/gennylib/include/gennylib/v1/Topology.hpp#L61

If we used the above, we could change `OnlyRunInInstances` to instead be a `RunThroughTopologyVisitor` or something, constructing a topology visitor using a mapping between visitor names and visitor functions. That would be a lot more general, and re-use existing code.

To see the topology types in use, see here: https://github.com/mongodb/genny/blob/master/src/gennylib/src/quiesce.cpp

Definitely push back if you don't think that'd be worth it!

Reply:
@aggrand I have been trying to implement this using the Topology and TopologyVisitor, but I ran into issues with it. I filed  [TIG-4060](https://jira.mongodb.org/browse/TIG-4060). Given the amount of time we already spent with this ticket, we would like to leave it as it is for now. We could open a follow up ticket to change this to use the Topology, and also allow configuration on a per operation basis.

Reply:
Yeah a followup ticket sounds great!